File: app/src/test/java/com/github/cythara/PitchComparatorTest.java
package com.github.cythara;

import com.github.cythara.domain.PitchComparator;
import com.github.cythara.domain.PitchDifference;

import org.junit.Assert;
import org.junit.Test;

import java.util.HashMap;
import java.util.Map;

import static com.github.cythara.data.tuning.GuitarTuning.Pitch.*;
import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.Matchers.closeTo;

public class PitchComparatorTest {

    @Test
    public void retrieveNote() throws Exception {
        Map<Float, PitchDifference> expectations = new HashMap<>();
        expectations.put(20f, new PitchDifference(E2, -2451.3202694972874));
        expectations.put(500f, new PitchDifference(E4, 721.3071582323822));
        expectations.put(197.67f, new PitchDifference(G3, 14.705999652460953));
        expectations.put(128.415f, new PitchDifference(D3, -232.0232233030192));

        for (Float pitch : expectations.keySet()) {
            PitchDifference actual = PitchComparator.INSTANCE.retrieveNote(pitch);
            PitchDifference expected = expectations.get(pitch);

            Assert.assertThat(actual.closest, is(expected.closest));
            Assert.assertThat(actual.deviation, closeTo(expected.deviation, 0.001));
        }
    }
}

File: base/src/test/java/com/github/cythara/PitchComparatorTest.kt
package com.github.cythara

import com.github.cythara.data.tuning.ChromaticTuning
import com.github.cythara.domain.Note
import com.github.cythara.domain.PitchComparator
import com.github.cythara.domain.PitchDifference
import org.hamcrest.CoreMatchers.`is`
import org.hamcrest.Matchers.closeTo
import org.junit.Assert
import org.junit.Test
import java.util.*

class PitchComparatorTest {

    @Test
    @Throws(Exception::class)
    fun retrieveNote() {
        val expectations = HashMap<Float, PitchDifference>()
        expectations[82.40689f] = PitchDifference(Note.createNote("E2"), 0.0)
        expectations[21f] = PitchDifference(Note.createNote("E2"), 33.149314659801846)
        expectations[500f] = PitchDifference(Note.createNote("B4"), 21.309458914380304)
        val tuning = ChromaticTuning()

        for (pitch in expectations.keys) {
            val (closest, deviation) = PitchComparator.retrieveNote(tuning, pitch)
            val expected = expectations[pitch]!!

            Assert.assertThat(closest.name, `is`(expected.closest.name))
            Assert.assertThat(deviation, closeTo(expected.deviation, 0.001))
        }
    }
}
